### PR TITLE
feat(core): Add Redis Hashes to cache service

### PR DESCRIPTION
**Pull request by flipswitchingmonkey**
cache service now supports storing values as Redis Hashes. if memory cache is used, these are stored as objects.

### CODE REVIEW

**Review by ivov — COMMENTED** [review]
Thanks for taking this on!

File: packages/cli/src/services/cache.service.ts [hunk 1]
```diff
@@ -106,6 +106,74 @@ export class CacheService extends EventEmitter {
...
+	 * If the cache is disabled, refreshFunction's result or fallbackValue is returned.
+	 *
+	 * If cache is not hit, and neither refreshFunction nor fallbackValue are provided, `undefined` is returned.
+	 * @param key The key to fetch from the cache
+	 * @param field Specific field to fetch from the hash
+	 * @returns value of the field in the hash
+	 */
+	async getHashField(key: string, field: string): Promise<unknown> {
+		return this.getHash(key, { fieldName: field });
+	}
```
[the remark this answers is not in the document]
Comment (flipswitchingmonkey) [reply packages/cli/src/services/cache.service.ts:121]: No idea, it's just for convenience.

File: packages/cli/src/services/cache.service.ts [hunk 2]
```diff
@@ -44,7 +44,7 @@ export class CacheService extends EventEmitter {
 			backend === 'redis' ||
 			(backend === 'auto' && config.getEnv('executions.mode') === 'queue')
 		) {
-			const { redisInsStore } = await import('cache-manager-ioredis-yet');
+			const { redisInsStore } = await import('./cache/cacheManagerRedis');
```
[the remark this answers is not in the document]
Comment (flipswitchingmonkey) [reply packages/cli/src/services/cache.service.ts:47]: you mean `await import('@/services/cache/cacheManagerRedis');`?

File: packages/cli/src/services/cache/cacheManagerRedis.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+		async hexists(key: string, field: string): Promise<boolean> {
+			return (await redisCache.hexists(key, field)) === 1;
+		},
+		async hdel(key: string, field: string) {
+			return redisCache.hdel(key, field);
+		},
+	} as RedisStore;
+}
+
+export function redisInsStore(redisCache: Redis | Cluster, options?: Config) {
```
[the remark this answers is not in the document]
Comment (flipswitchingmonkey) [reply packages/cli/src/services/cache/cacheManagerRedis.ts:150]: Taken from the original code. I don't think it's referenced anywhere else, so we can rename it to something better.

File: packages/cli/src/services/cache.service.ts [hunk 4]
```diff
@@ -106,6 +106,74 @@ export class CacheService extends EventEmitter {
...
+	 * @returns an object with the key/value pairs of the hash or a single value if fieldName is provided
+	 */
+	async getHash<T>(
+		key: string,
+		options: {
+			fieldName?: string;
+			fallbackValue?: T | Record<string, unknown>;
+			refreshFunction?: (key: string) => Promise<unknown>;
+		} = {},
+	): Promise<Record<string, unknown> | T | undefined> {
```
[the remark this answers is not in the document]
Comment (flipswitchingmonkey) [reply packages/cli/src/services/cache.service.ts:142]: i remember the decision some time back being not to enforce having to set the type for the caches

File: packages/cli/src/services/cache.service.ts [hunk 5]
```diff
@@ -106,6 +106,74 @@ export class CacheService extends EventEmitter {
...
+		if (!key || key.length === 0) {
+			return;
+		}
```
[the remark this answers is not in the document]
Comment (flipswitchingmonkey) [reply packages/cli/src/services/cache.service.ts:145]: 1. i think those are all leftovers from a copy & paste
2. how could you set the cache without knowing having a keyname for it?

File: packages/cli/src/services/cache.service.ts [hunk 6]
```diff
@@ -305,4 +445,40 @@ export class CacheService extends EventEmitter {
...
+	/**
+	 * Return all keys in a cached hash.
+	 */
+	async hashValues<T>(key: string): Promise<T[]> {
+		if (this.isRedisCache()) {
+			return (this.cache?.store as RedisStore).hvals<T>(key);
+		} else {
+			const obj = (await this.cache?.store.get(key)) ?? {};
+			return Object.values(obj) as T[];
+		}
```
[the remark this answers is not in the document]
Comment (flipswitchingmonkey) [reply packages/cli/src/services/cache.service.ts:470]: tbh i prefer the explicit else, but we can do that

File: packages/cli/src/services/cache.service.ts [hunk 7]
```diff
@@ -305,4 +445,40 @@ export class CacheService extends EventEmitter {
...
+		} else {
+			const obj = (await this.cache?.store.get(key)) ?? {};
+			return Object.keys(obj);
+		}
+	}
+
+	/**
+	 * Return all keys in a cached hash.
+	 */
+	async hashValues<T>(key: string): Promise<T[]> {
```
[the remark this answers is not in the document]
Comment (flipswitchingmonkey) [reply packages/cli/src/services/cache.service.ts:464]: faulty comment, returns values, not keys

File: packages/cli/src/services/cache.service.ts [hunk 8]
```diff
@@ -2,7 +2,7 @@ import { Service } from 'typedi';
 import config from '@/config';
 import { caching } from 'cache-manager';
 import type { MemoryCache } from 'cache-manager';
-import type { RedisCache } from 'cache-manager-ioredis-yet';
+import type { RedisCache, RedisStore } from './cache/cacheManagerRedis';
```
Comment (ivov) [inline packages/cli/src/services/cache.service.ts:5]: Ideally absolute import - we need a lint rule for this.

File: packages/cli/src/services/cache.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivov) [reply packages/cli/src/services/cache.service.ts:160]: In that case you can pass in a default `<T = unknown>`

File: packages/cli/src/services/cache.service.ts [hunk 9]
```diff
@@ -127,10 +225,28 @@ export class CacheService extends EventEmitter {
...
 	): Promise<unknown[]> {
-		if (keys.length === 0) {
-			return [];
+		if (!this.isEnabled) {
+			if (options.refreshFunctionMany) {
+				return options.refreshFunctionMany(keys);
+			}
+			if (options.refreshFunctionEach) {
+				const values = await Promise.all(
+					keys.map(async (key) => options.refreshFunctionEach!(key)),
```
Comment (ivov) [inline packages/cli/src/services/cache.service.ts:234]: Odd that `!` is needed since it's being checked above.

**Review by ivov — APPROVED** [review]
Had to do a rather quick pass as I need to continue with RBAC - please note there remain some open comments, we can either address or look into them after merging, up to you.

**Comment by ivov** [comment]
Continued at #8213

**Pull request closed without merging**